aesd-char-driver/main.c

```diff
@@ -0,0 +1,379 @@
+	{
+		PDEBUG("Error in read mutex locking");
+		return -ERESTARTSYS; //Allow system to be restartable
+	}
+
+	new_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&char_dev->buffer, *f_pos, &byte_offset);
+
+	if(new_entry)
+	{
+		size_t bytes_remaining = new_entry->size - byte_offset;
+		size_t bytes_to_copy = min(bytes_remaining, count);
+        
+        //copy data from kernel space to user space and check for number of bytes that could not be copied
+
+		if (copy_to_user(buf, new_entry->buffptr + byte_offset, bytes_to_copy)) 
+		{
+			retval = -EFAULT;
+		} 
+		else 
+		{
+			retval = bytes_to_copy;
+			*f_pos += retval;
+		}
+	}
+
+	// Unlock the mutex
+	mutex_unlock(&char_dev->lock);
+
+	return retval;
+}
+
+/**
+ * @brief Write data to the AESD device.
+ * 
+ * The function handles writing data from a user space buffer to the AESD device. If a newline character 
+ * is encountered, the data up to and including the newline is added to a device buffer and, if filled, 
+ * the device buffer data is added to a circular buffer. This function uses a mutex to ensure mutual 
+ * exclusion and ensures synchronization across potential concurrent writes. 
+ * 
+ * @param filp A pointer to the file structure representing the device file.
+ * @param buf A user space buffer containing the data to write.
+ * @param count The number of bytes to write from the user space buffer.
+ * @param f_pos A pointer to the file position variable. This will be updated based on the number of bytes written.
+ * 
+ * @return Returns the number of bytes successfully written to the device. On failure, it returns an appropriate error code.
+ */
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos)
+{
+    // Initialization of return value to a memory error code.
+    ssize_t retval = -ENOMEM;
+
+    // Check if either the file pointer 'filp' or the buffer 'buf' is NULL.
+    if((!filp) || (!buf))
+    {
+        return -EINVAL;
+    }
+
+    // Debug message to log the size of data and offset being written.
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+
+    // Extracting character device instance from the file structure.
+    struct aesd_dev *char_dev = filp->private_data;
+	// char_dev->write_buffer = NULL;
+	// char_dev->write_buffer_size = 0;
+
+    // Dynamically allocate memory for the data to be written.
+    char *write_data = kmalloc(count, GFP_KERNEL);
+
+    // Check for memory allocation failure.
+    if(!write_data)
+    {
+        //Handle error case
+        PDEBUG("Kmalloc failed while writing");
+        return retval;
+    }
+
+    // Copy data from user space to kernel space.
+    if (copy_from_user(write_data, buf, count)) 
+    {
+        //Handler error case
+        retval = -EFAULT;
+        kfree(write_data);
+        return retval;
+    }
+
+    // Lock the mutex for synchronizing access.
+    if(mutex_lock_interruptible(&char_dev->lock) != 0)
```

Comment:
@Suraj-Ajjampur you could move this to line 224 I believe

Reply:
Change is made with the latest commit.